src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { PrincipalType } from '../../../../core/utils/constants';
+
+export interface Principals {
+  users?: string[];
+  groups?: string[];
+}
+
+export type Permissions = Partial<Record<string, Principals>>;
+
+const addPermissionToPrincipals = (
+  principals?: Principals,
+  users?: string[],
+  groups?: string[]
+) => {
+  if (!principals) {
+    principals = {};
+  }
+  if (!!users) {
+    if (!principals.users) {
+      principals.users = [];
+    }
+    principals.users = Array.from(new Set([...principals.users, ...users]));
+  }
+  if (!!groups) {
+    if (!principals.groups) {
+      principals.groups = [];
+    }
+    principals.groups = Array.from(new Set([...principals.groups, ...groups]));
+  }
+  return principals;
+};
+
+const deletePermissionFromPrincipals = (
+  principals?: Principals,
+  users?: string[],
+  groups?: string[]
+) => {
+  if (!principals) {
+    return principals;
+  }
+  if (!!users && !!principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (!!groups && !!principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+export class ACL {
+  private permissions?: Permissions;
+  constructor(initialPermissions?: Permissions) {
+    this.permissions = initialPermissions;
+  }
+
+  // parse the permissions object to check whether the specific user or group has the specific permission or not
+  public hasPermission(
+    permissionType: string,
+    permissions: Permissions,
+    user?: string,
+    group?: string
+  ) {
+    if ((!user && !group) || !permissionType || !permissions) {
+      return false;
+    }
+
+    const principals = permissions[permissionType];
+    if (!!principals) {
+      if (
+        !!user &&
+        !!principals.users &&
+        (principals.users.includes('*') || principals.users.includes(user))
+      ) {
+        return true;
+      }
+      if (
+        !!group &&
+        !!principals.groups &&
+        (principals.groups.includes('*') || principals.groups.includes(group))
+      ) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  // permissions object build function, add users or groups with specific permission to the object
+  public addPermission(permissionTypes: string[], users?: string[], groups?: string[]) {
+    if ((!users && !groups) || !permissionTypes) {
+      return this;
+    }
+    if (!this.permissions) {
+      this.permissions = {};
+    }
```

Comment:
how about moving this to constructor? so that we don't need to repeat it in `removePermission` method